File: src/main/python/leetcode_by_python/SegmentOne/No10_Regular_Expression_Matching.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019/4/8 21:06
# https://leetcode.com/problems/regular-expression-matching/


class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        lenS, lenP = len(s), len(p)
        dp = [[False] * (lenP + 1) for _ in range(lenS + 1)]

        dp[0][0] = True
        for i in range(1, lenP):
            dp[0][i + 1] = dp[0][i - 1] and p[i] == '*'

        for i in range(lenS):
            for j in range(lenP):
                if p[j] == s[i] or p[j] == '.':
                    dp[i + 1][j + 1] = dp[i][j]
                if p[j] == '*':
                    if p[j - 1] != s[i] and p[j - 1] != '.':
                        dp[i + 1][j + 1] = dp[i + 1][j - 1]
                    else:
                        dp[i + 1][j + 1] = dp[i][j + 1] or dp[i + 1][j] or dp[i + 1][j - 1]

        return dp[lenS][lenP]


if __name__ == '__main__':
    solution = Solution()
    assert solution.isMatch('aa', 'a') == False
    assert solution.isMatch('aa', 'a*') == True
    assert solution.isMatch('ab', '.*') == True
    assert solution.isMatch('aab', 'c*a*b') == True
    assert solution.isMatch('mississippi', 'mis*is*p*.') == False

File: src/main/java/com/hazza/leetcode/SegmentOne/No10RegularExpressionMatching.java
package com.hazza.leetcode.SegmentOne;

/**
 * https://leetcode.com/problems/regular-expression-matching/#/description
 */
public class No10RegularExpressionMatching {
    public boolean isMatch(String s, String p) {
        if (s.equals("") || s == null || p.equals("") || p == null) return false;
        if (s.equals(p)) return true;
        int lenS = s.length(), lenP = p.length();
        int start = -1;

        for (int i = 0; i < lenP; i++) {
            if (s.charAt(0) == p.charAt(i) || p.charAt(i) == '.') {
                start = i + 1;
                break;
            }
        }
        if (start == -1 || (start == lenP && lenS > 1)) return false;

        for (int i = 1; i < lenS; i++) {
            if (start == lenP) return false;
            if (s.charAt(i) == p.charAt(start) || p.charAt(start) == '.') {
                ++start;
            } else if (p.charAt(start) == '*') {
                if (p.charAt(start - 1) == '.') return true;
                if (p.charAt(start - 1) != s.charAt(i)) ++start;
            }  else return false;
        }
        return true;
    }
}
